policy/release/rpm_ostree_task.rego

```diff
@@ -0,0 +1,153 @@
+#
+deny contains result if {
+	some error in builder_image_param_errors
+	result := _with_effective_on(lib.result_helper(rego.metadata.chain(), [error.msg]), error)
+}
+
+# METADATA
+# title: Rule data
+# description: >-
+#   Verify the rule data used by this package, `allowed_rpm_ostree_builder_image_prefixes`, is in
+#   the expected format.
+# custom:
+#   short_name: rule_data
+#   failure_msg: "%s"
+#   solution: >-
+#     Make sure the `allowed_rpm_ostree_builder_image_prefixes` rule data is in the expected format
+#     in the data source.
+#   collections:
+#   - redhat
+#
+deny contains result if {
+	some error in rule_data_errors
+	result := lib.result_helper(rego.metadata.chain(), [error])
+}
+
+# Detect when an image reference is not pinned to a digest.
+builder_image_param_errors contains error if {
+	some image in _builder_images
+	not contains(image, "@sha256:")
+	error := {"msg": sprintf("BUILDER_IMAGE %q is not pinned to a digest", [image])}
+}
+
+# Detect when an image reference does not start with with any of the pre-approved prefixes.
+builder_image_param_errors contains error if {
+	some image in _builder_images
+
+	# There are no matches
+	count([prefix |
+		some prefix in _allowed_prefixes
+		startswith(image, prefix.value)
+	]) == 0
+
+	pretty_prefixes := concat(", ", [prefix.value | some prefix in _allowed_prefixes])
+
+	error := {"msg": sprintf(
+		"BUILDER_IMAGE %q does not start with a pre-approved prefix: %s",
+		[image, pretty_prefixes],
+	)}
+}
+
+# Detect when an image starts with a pre-approved prefix, but that prefix has expiration date.
+builder_image_param_errors contains error if {
+	some image in _builder_images
+
+	# There is a match, but it has an expiration date.
+	some prefix in _allowed_prefixes
+	startswith(image, prefix.value)
+	prefix.expires_on != ""
+
+	error := {
+		"msg": sprintf(
+			"BUILDER_IMAGE %q starts with %q prefix that expires on %s",
+			[image, prefix.value, prefix.expires_on],
+		),
+		"effective_on": prefix.expires_on,
+	}
+}
+
+rule_data_errors contains msg if {
+	schema := {
+		"$schema": "http://json-schema.org/draft-07/schema#",
+		"type": "array",
+		"items": {"anyOf": [
+			{
+				"type": "object",
+				"properties": {"value": {"type": "string"}, "expires_on": {"type": "string"}},
+				"additionalProperties": false,
+				"required": ["value"],
+			},
+			{"type": "string"},
+		]},
+		"uniqueItems": true,
+	}
+
+	# match_schema expects either a marshaled JSON resource (String) or an Object. It doesn't
+	# handle an Array directly.
+	value := json.marshal(lib.rule_data(_rule_data_key))
+	some violation in json.match_schema(value, schema)[1]
+
+	# When the item doesn't match any of the expected formats, two errors are produced. One contains
+	# precise information of what is missing while the other, the "anyOf" error, is generic enough
+	# to be mostly useless. To avoid noise, the "anyOf" errors are excluded.
+	not violation.type == "number_any_of"
```

Comment:
I'd include both, I think folk can handle the noise a bit at the cost of accidentally missing the important error...

Reply:
Good point. Done.